### PR TITLE
Edit sync

### LINKED ISSUE #299: edit sync
As a User
in order to adjust a sync to changing needs
it should be possible to change (the config of) an existing sync

### PR DESCRIPTION
**Pull request by julianseeger**
resolves #299 edit sync

+ Introduces Cucumber for high-level tests
+ a little bit polluted because I had to stabilize the gui tests very hard after updating to u16.04

### CODE REVIEW

File: src/test-gui/java/de/qabel/desktop/ui/AScreenshotOnFailureWatcher.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    protected void failed(Throwable e, Description description) {
+        AsyncUtils.runLaterAndWait(() -> {
+            Window window = robot.targetWindow();
+            double x = window.getX();
+            double y = window.getY();
+            double width = window.getWidth();
+            double height = window.getHeight();
+            System.out.println("capturing " + window.getScene());
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/AScreenshotOnFailureWatcher.java]: println?

File: src/test-gui/java/de/qabel/desktop/ui/AScreenshotOnFailureWatcher.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    new File(
+                        "build/reports/guiTest/screenshots/"
+                            + description.getClassName()
+                            + "."
+                            + description.getMethodName()
+                            + ".png"
+                    )
+                );
+            } catch (Exception e1) {
+                e1.printStackTrace();
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/AScreenshotOnFailureWatcher.java]: meh

File: src/test-gui/java/de/qabel/desktop/ui/AbstractControllerGUITest.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
 import static de.qabel.desktop.AsyncUtils.waitUntil;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/AbstractControllerGUITest.java]: unused import

File: src/test-gui/java/de/qabel/desktop/ui/AbstractGuiTest.java [hunk 4]
```diff
@@ -59,6 +67,11 @@ protected Object launchNode(FXMLView view) {
         baseFXRobot = new BaseFXRobot(scene);
         baseFXRobot.waitForIdle();
         waitTillTheEnd(robot.rootNode(scene));
+        baseFXRobot.waitForIdle();
+        /*runLaterAndWait(() -> {
+            robot.targetWindow(stage);
+            stage.toFront();
+        });*/
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/AbstractGuiTest.java]: commented code

File: src/test-gui/java/de/qabel/desktop/ui/AbstractPage.java [hunk 5]
```diff
@@ -27,7 +32,76 @@ public AbstractPage(FXRobot baseFXRobot, FxRobot robot) {
...
+            false,
+            false,
+            new PickResult(node, pos.getX(), pos.getY())
+        );
+        EventHandler<? super MouseEvent> pressedHandler = node.getOnMousePressed();
+        EventHandler<? super MouseEvent> clickedHandler = node.getOnMouseClicked();
+        EventHandler<? super MouseEvent> releasedHandler = node.getOnMouseReleased();
+
+        Platform.runLater(() -> {
+            //node.buildEventDispatchChain(new EventDispatchChainImpl()).dispatchEvent(event);
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/AbstractPage.java]: commented code

File: src/test-gui/java/de/qabel/desktop/ui/PlatformTearDownWatcher.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    private void tearDown() {
+        if (stage != null) {
+            Platform.runLater(() -> { try { stage.close(); } catch (Exception ignored) {}});
+        }
+        /*for (Window window : robot.robotContext().getWindowFinder().listTargetWindows()) {
+            AsyncUtils.runLaterAndWait(window::hide);
+        }
+        stage = null;
+        System.gc();*/
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/PlatformTearDownWatcher.java]: more commented code

File: src/test-gui/java/de/qabel/desktop/ui/sync/SyncStepdefs.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    @And("^'(.+)' exists in remote '(.+)'$")
+    public void fileExistsInRemote(String filename, String remotePath) throws Throwable {
+        Path remote = BoxFileSystem.get("/", remotePath);
+        final BoxNavigation remoteNavigation = navigate(remote);
+        waitUntil(() -> remoteNavigation.hasFile(filename));
+    }
+
+    private BoxNavigation navigate(Path remote) throws QblStorageException {
+        BoxNavigation navigation = volume.navigate();
+        for (int i = 0; i < remote.getNameCount(); i++) {
```
Comment (audax) [inline src/test-gui/java/de/qabel/desktop/ui/sync/SyncStepdefs.java]: Is there no way to just iterate through the names directly?

File: src/test/java/de/qabel/desktop/AsyncUtils.java [hunk 8]
```diff
@@ -32,4 +34,36 @@ public static void waitUntil(Callable<Boolean> evaluate, long timeout, Callable<
...
+                Thread.yield();
+                Thread.sleep(10);
+                assertion.run();
+                return;
+            } catch(AssertionError e){
+                if (System.currentTimeMillis() - timeout > startTime) {
+                    throw e;
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (audax) [inline src/test/java/de/qabel/desktop/AsyncUtils.java]: meh

Comment (julianseeger) [inline src/test/java/de/qabel/desktop/AsyncUtils.java]: The fail only preserves the message. So we need a Stacktrace from anywhere. And in test code I see no point in roundtripping through a logger, do you?

File: src/test-gui/java/de/qabel/desktop/ui/AScreenshotOnFailureWatcher.java (hunk 2 above)
Comment (julianseeger) [inline src/test-gui/java/de/qabel/desktop/ui/AScreenshotOnFailureWatcher.java]: what else? The test failed, we tried to at least capture a screenshot but even that didn't work ... now we leave a trace why it failed and let the tests continue...

File: src/test-gui/java/de/qabel/desktop/ui/sync/SyncStepdefs.java (hunk 7 above)
Comment (julianseeger) [inline src/test-gui/java/de/qabel/desktop/ui/sync/SyncStepdefs.java]: Didn't know it yet but I just found out that Path implements Iterable .. so looping over that path itself will provide its names. Going to fix that...

**Pull request merged**